fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java

```diff
@@ -0,0 +1,458 @@
+    // ColumnRefRemapping is used to keep the mapping: old slotId->new slotId
+    public static class ColumnRefRemapping {
+        private final Map<ColumnRefOperator, ColumnRefOperator> remapping;
+        private Optional<ReplaceColumnRefRewriter> cachedReplacer = Optional.empty();
+        private Optional<ColumnRefSet> cachedColumnRefSet = Optional.empty();
+        public static final ColumnRefRemapping EMPTY_REMAPPING = new ColumnRefRemapping();
+
+        public ColumnRefRemapping() {
+            remapping = Maps.newHashMap();
+        }
+
+        public ColumnRefRemapping(Map<ColumnRefOperator, ColumnRefOperator> remapping) {
+            this.remapping = remapping;
+        }
+
+        void combine(ColumnRefRemapping other) {
+            remapping.putAll(other.remapping);
+            cachedReplacer = Optional.empty();
+            cachedColumnRefSet = Optional.empty();
+        }
+
+        ReplaceColumnRefRewriter getReplacer() {
+            if (!cachedReplacer.isPresent()) {
+                ReplaceColumnRefRewriter replacer =
+                        new ReplaceColumnRefRewriter(remapping.entrySet().stream().collect(
+                                Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)), true);
+                cachedReplacer = Optional.of(replacer);
+            }
+            return cachedReplacer.get();
+        }
+
+        ColumnRefSet getColumnRefSet() {
+            if (!cachedColumnRefSet.isPresent()) {
+                cachedColumnRefSet = Optional.of(new ColumnRefSet(remapping.keySet()));
+            }
+            return cachedColumnRefSet.get();
+        }
+
+        boolean isEmpty() {
+            return remapping.isEmpty();
+        }
+    }
+
+    public static class RewriteInfo {
+        private boolean rewritten = false;
+        private ColumnRefRemapping remapping;
+        private OptExpression op;
+
+        public AggregatePushDownContext getCtx() {
+            return ctx;
+        }
+
+        public void setCtx(AggregatePushDownContext ctx) {
+            this.ctx = ctx;
+        }
+
+        private AggregatePushDownContext ctx;
+
+        public static RewriteInfo NOT_REWRITE = new RewriteInfo(false, null, null, null);
+
+        public RewriteInfo(boolean rewritten, ColumnRefRemapping remapping,
+                           OptExpression op, AggregatePushDownContext ctx) {
+            this.rewritten = rewritten;
+            this.remapping = remapping;
+            this.op = op;
+            this.ctx = ctx;
+        }
+
+        public boolean hasRewritten() {
+            return rewritten;
+        }
+
+        public void setRewritten(boolean rewritten) {
+            this.rewritten = rewritten;
+        }
+
+        public Optional<ColumnRefRemapping> getRemapping() {
+            if (!rewritten || remapping.isEmpty()) {
+                return Optional.empty();
+            }
+            return Optional.of(remapping);
+        }
```

Comment:
I think don't need Optional method and may be easier if add a method `hasRemapping`?

Reply:
Optional object is more convenient, we can invoke orElse, ifPresent, map method.